appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,725 @@
+  var mate = Genetics.Cage.miceMap_[chosen.id];
+  if (Genetics.Cage.isMatingSuccessful(mouse, mate)) {
+    // Create offspring from the two mice.
+    Genetics.Cage.createOffspring(mouse, mate);
+    // Account for overpopulation.
+    if (Genetics.Cage.aliveMice_.length > Genetics.Cage.MAX_POPULATION) {
+      // Find oldest mouse.
+      var oldestMouse = Genetics.Cage.aliveMice_[0];
+      for (var i = 1; i < Genetics.Cage.aliveMice_.length; i++) {
+        var aliveMouse = Genetics.Cage.aliveMice_[i];
+        if (!oldestMouse || aliveMouse.age > oldestMouse.age) {
+          oldestMouse = aliveMouse;
+        }
+      }
+      // Kill oldest mouse.
+      new Genetics.Cage.Event('OVERPOPULATION', oldestMouse.id).addToQueue();
+      Genetics.Cage.death(oldestMouse);
+    }
+  } else {
+    // Use up one mating attempt for mouse.
+    mouse.fertility--;
+  }
+};
+
+/**
+ * Creates a child mouse from two mice and adds offspring to the cage.
+ * @param {!Genetics.Mouse} parent1 One of the parents of the new mouse.
+ * @param {!Genetics.Mouse} parent2 One of the parents of the new mouse.
+ */
+Genetics.Cage.createOffspring = function(parent1, parent2) {
+  // Use up one mating attempt for each of the parents.
+  parent1.fertility--;
+  parent2.fertility--;
+  // Allocate a new id for the mouse.
+  var mouseId = Genetics.Cage.nextAvailableMouseId_++;
+  // Determine sex of child based on the current population.
+  var populationFertility = 0;
+  var femaleFertility = 0;
+  for (var i = 0; i < Genetics.Cage.aliveMice_.length; i++) {
+    var aliveMouse = Genetics.Cage.aliveMice_[i];
+    populationFertility += aliveMouse.fertility;
+    if (aliveMouse.sex == Genetics.Mouse.Sex.FEMALE) {
+      femaleFertility += aliveMouse.fertility;
+    }
+  }
+  var childSex = ( Math.random() < femaleFertility / populationFertility) ?
+      Genetics.Mouse.Sex.MALE : Genetics.Mouse.Sex.FEMALE;
+  var child = new Genetics.Mouse(mouseId, childSex, parent1, parent2);
+  Genetics.Cage.born(child);
+  new Genetics.Cage.Event('MATE', parent1.id, 'SUCCESS', parent2.id,
+      child).addToQueue();
+};
+
+/**
+ * Checks whether a mate between the two mice will succeed.
+ * @param {Genetics.Mouse} proposingMouse The mouse that is proposing to mate.
+ * @param {Genetics.Mouse} askedMouse The mouse that is being asked to mate.
+ * @return {boolean} Whether the mate succeeds.
+ */
+Genetics.Cage.isMatingSuccessful = function(proposingMouse, askedMouse) {
+  // Check if mice are the same.
+  if (proposingMouse.id == askedMouse.id) {
+    new Genetics.Cage.Event('MATE', proposingMouse.id, 'SELF').addToQueue();
+    return false;
+  }
+  // Check if mice are not of compatible sex.
+  if (proposingMouse.sex != Genetics.Mouse.Sex.HERMAPHRODITE &&
+      proposingMouse.sex == askedMouse.sex) {
+    new Genetics.Cage.Event('MATE', proposingMouse.id, 'INCOMPATIBLE',
+        askedMouse.id).addToQueue();
+    return false;
+  }
+  // Check if other mouse is infertile.
```

Comment:
proposing mouse is guaranteed to be fertile because this function is called when a mouse (the proposing mouse) is fertile.